### DISCUSSION TITLE
Trigger Listerner being overwritten no idea why

### DISCUSSION THREAD

**Discussion by CrazyMoka** [quartznet/quartznet#1757] [Q&A] [answered]
Hello,
In this scenario the event listener for triggers are being removed.

Both triggers are running the same durrble job with DisallowConcurrentExecution. 

Trigger A is established to run every minute and log stuff to the DB through the event listener. Trigger has a cron with a start time No end time.

Trigger B is initiated with a start time 5 minutes after trigger B and scheduled to run with a cron everyt 1 minute for 5 minutes.

Now. Trigger A is blocked by Trigger B even though the start time for B is well passed trigger A.
Trigger B starts running and logs info to the DB. When it ends trigger A starts working but the even listerner doesn't get called.

It is as if Trigger A listerner was removed when it was blocked by Trigger B?

No sure what is happening here.

Please advise.

**Comment by lahma** [comment]
Is it being removed, is there some log information?

**Reply by CrazyMoka** [reply]
I know its fosure not being fired because the job logs are printing to the screen but the logs from the listerner are not.

**Reply by lahma** [reply]
Maybe you can share a repro as GitHub repository or something similar so we don't need to guess so much?

**Reply by CrazyMoka** [reply]
https://pastebin.com/fMe3XSuD

line 231 is where I attach the triggers when they are create.

**Reply by CrazyMoka** [reply]
Forgot to show
in the Job code.
				logger.Information(context.JobDetail.Key + " job executing, triggered by " + context.Trigger.Key);
				logger.Information("Meesage " + message);
				logger.Information("user_id " + user_id);
				
				This gets show in the console when the job is running , however the logs from the listerner events do not show any log information. Therefor they are not being run.
				
				No idea why its "removed" or how to tell if it indeed is being removed.

**Comment by CrazyMoka** [comment]
I think the issue is:
```
            // attach the trigger to the listener system. This helps with logging events and capturing errors.
            UpdateTaskDetailLogListener _taskListener = new UpdateTaskDetailLogListener(id);

            IMatcher<TriggerKey> matcher = KeyMatcher<TriggerKey>.KeyEquals(_trigger.Key);

            scheduler.ListenerManager.AddTriggerListener(_taskListener, matcher);

            _logger.Information("----- GETTING CURRENT LISTENERS AFTER CREATE --------");
            foreach (var _l in scheduler.ListenerManager.GetTriggerListeners())
            {

                _logger.Information(_l.ToString());

            }
```
As looping through the Listerner Manager for triggers always only returns one. So it must be overwriting it somehow.
Tigger Keys are all different as they are a random string.
The Id is always 1 for these particular jobs. So is it makien the same object event though its NEW ?

Regards

**Comment by CrazyMoka** [comment]
Another Update:
This runs on startup from the schedule listerer to attach the Trigger Listers
```
            //TODO check to see if trigger has a log if it doesn't remove the trigger.

            foreach (TaskDetailLog item in unfinishedJobs)
            {
                

                var _triggerKey = new TriggerKey(item.TriggerName, item.TriggerGroup);
                UpdateTaskDetailLogListener _taskListener = new UpdateTaskDetailLogListener(item.TaskDetailId, _triggerKey.ToString());

                IMatcher<TriggerKey> matcher = KeyMatcher<TriggerKey>.KeyEquals(_triggerKey);

                scheduler.ListenerManager.AddTriggerListener(_taskListener, matcher);

                logger.Information("--- Sched Listener --- Trigger listener added for " + item.TriggerName + "." + item.TriggerGroup);
            }

```
This is what happens
---- TRIGGER Created for ------1
[15:01:48 INF] --- Sched Listener --- Trigger listener added for Trigger_AIUCJY.Scheduled
---- TRIGGER Created for ------1
[15:01:48 INF] --- Sched Listener --- Trigger listener added for Trigger_M9O9C3.Scheduled
---- TRIGGER Created for ------1
[15:01:48 INF] --- Sched Listener --- Trigger listener added for Trigger_JF86F8.Scheduled
[15:01:48 INF] ----- total of 1 Trigger Listeners --------

As you can see 
            logger.Information("----- total of " + scheduler.ListenerManager.GetTriggerListeners().Count + " Trigger Listeners --------");

Only one trigger listener is added so they are all being over written.

Does this mean only one trigger can belong to one trigger listerner?

Or am I not assigning my trigger listerners properly?

**Comment by CrazyMoka** [comment, marked as answer]
NAme has to be unique to each trigger listerner object.  
  public override string Name => "1st Trigger Listener: "+triggerKey;

**Reply by CrazyMoka** [reply]
This should be in the documentation somewhere... If it is it needs to be more prominent.
I found the answer in stack overflow for Job Event Listener with the same issue.

**Reply by lahma** [reply]
There's the docs folder if you want to create PRs and improve, there's also link in the quartz-scheduler.net web page on each page that allows you to submit PRs to improve specific pages.

**Reply by CrazyMoka** [reply]
Thank you! I'll try to help improve docs. As I will be using Quartz for this project.
